docs/types/operators.rst

```diff
@@ -69,7 +69,7 @@ delete
 ``delete a`` assigns the initial value for the type to ``a``. I.e. for integers it is
 equivalent to ``a = 0``, but it can also be used on arrays, where it assigns a dynamic
 array of length zero or a static array of the same length with all elements set to their
-initial value. ``delete a[x]`` deletes the item at index ``x`` of the array and leaves
+initial value, calling `delete` on each array element, both for static and dynamic arrays. ``delete a[x]`` deletes the item at index ``x`` of the array and leaves
```

Comment:
This is not true. It's worded the way it is exactly because it is not necessarily equivalent to recursively calling `delete` on every element. For dynamic arrays the effect is equivalent to assigning an array of length zero - only the length changes and the elements stay unchanged. They will only be zeroed when you enlarge the array, e.g. with `push()`.

Reply:
> For dynamic arrays the effect is equivalent to assigning an array of length zero - only the length changes and the elements stay unchanged

But it does get cleared: 

https://github.com/ethereum/solidity/blob/983407762c3423e9c301d5ae56ac7b6d951655df/libsolidity/codegen/YulUtilFunctions.cpp#L1759
https://github.com/ethereum/solidity/blob/983407762c3423e9c301d5ae56ac7b6d951655df/libsolidity/codegen/YulUtilFunctions.cpp#L1727-L1730

Reply:
> We don't clear the slot when doing a `.push()`

Ah, right. That was a part of that bug we had some time ago. I constantly forget that because it's kinda unintuitive.

> But it does get cleared:

Hmm... ok, does that work the same for memory though? I remember it working like this somewhere, but maybe I remember it wrong.

Does assigning an empty array always match deleting stuff recursively?

Reply:
I just did a quick check with `solc --ir` and looks like, at the very least in some cases in the new codegen`, `delete` does not work like this for memory. The array is simply assigned the location of the zero slot:

```yul
/// @src 0:122:130  "delete m"
let _5_mpos := zero_value_for_split_t_array$_t_uint256_$dyn_memory_ptr()
var_m_3_mpos := _5_mpos
```
```yul
function zero_value_for_split_t_array$_t_uint256_$dyn_memory_ptr() -> ret {
    ret := 96
}
```

So the text still is not entirely correct.